documentation/asciidoc/accessories/camera/libcamera_apps_post_processing.adoc

```diff
@@ -0,0 +1,216 @@
+=== Post-Processing
+
+`libcamera-apps` share a common post-processing framework. This allows them to pass the images received from the camera system through a number of custom image processing and image analysis routines. Each such routine is known as a _post-processing stage_ and the description of exactly which stages should be run, and what configuration they may have, is supplied in a JSON file. Every stage, along with its source code, is supplied with a short example JSON file showing how to enable it.
+
+For example, the simple _negate_ stage (which "negates" all the pixels in an image, turning light pixels dark and vice versa) is supplied with a `negate.json` file that configures the post-processing pipeline to run it:
+
+`libcamera-hello --post-process-file /path/to/negate.json`
+
+The negate stage is particularly trivial and has no configuration parameters of its own, therefore the JSON file merely has to name the stage, with no further information, and it will be run. Thus `negate.json` contains
+
+----
+{
+    "negate":
+    {
+    }
+}
+----
+
+To run multiple post-processing stages, the contents of the example JSON files merely need to be listed together, and the stages will be run in the order given. For example, to run the Sobel stage (which applies a Sobel filter to an image) followed by the negate stage we could create a custom JSON file containing
+
+----
+{
+    "sobel_cv":
+    {
+        "ksize": 5
+    },
+    "negate":
+    {
+    }
+}
+----
```

Comment:
This is effectively a "dictionary" structure (to use the Python naming) and in most programming languages a dictionary is unordered, so if a script was to read in this JSON file and then write it out again, I think it woud be perfectly valid for it to get written out as:
```
{
    "negate":
    {
    },
    "sobel_cv":
    {
        "ksize": 5
    }
}
```
which conflicts somewhat with your "the stages will be run in the order given" :confused: 
In strict "JSON syntax" if you want to preserve ordering, I think you'd need to use an array `[ ... ]` rather than a dictionary `{ ... }`.
(And apologies if this isn't the right place to be discussing this!)

Reply:
Hmm, interesting, I don't really have a good answer for that. I suppose the situation is that *our* implementation reads them in order, and that the order matters to us so I won't change this just for the moment. For reference, what would this look like using `[ ... ]`?

Reply:
Well if each of the steps has a unique name (which it should for a dictionary data-structure), then I guess you could perhaps do something like:
```json
{
    "config":
    {
        "sobel_cv":
        {
            "ksize": 5
        },
        "negate":
        {
        }
    },
    "order": ["sobel_cv", "negate"]
}
```
i.e. "config" is an unordered dictionary, and "order" specifies what order each of those operations should be performed in.

But if you'd ever want to do the same step in multiple places (perhaps with different options for each version of the same operation), i.e. something like this in your current syntax:
```json
{
    "sobel_cv":
    {
        "ksize": 5
    },
    "negate":
    {
    },
    "sobel_cv":
    {
        "ksize": 3
    }
}
```
then having a dictionary with multiple keys named the same is invalid JSON syntax (if I throw that example into https://jsonlint.com/ it complains with `Error: Duplicate key 'sobel_cv'`).

So In that case maybe you'd want to do something like:
```json
{
    "operations":
    [
        {
            "name": "sobel_cv",
            "config":
            {
                "ksize": 5
            }
        },
        {
            "name": "negate",
            "config":
            {
            }
        },
        {
            "name": "sobel_cv",
            "config":
            {
                "ksize": 3
            }
        }
    ]
}
````
i.e. an (ordered) array of dictionaries.
:man_shrugging: 

You _could_ get rid of the outer-dictionary i.e.
```json
[
    {
        "name": "sobel_cv",
        "config":
        {
            "ksize": 5
        }
    },
    {
        "name": "negate",
        "config":
        {
        }
    },
    {
        "name": "sobel_cv",
        "config":
        {
            "ksize": 3
        }
    }
]
```
but IMHO leaving it in (as in the prior example) gives you more flexibility for the future (i.e. you can add additional top-level dictionary keys without affecting the "operations" key, which you couldn't do if the top-level data structure was an array).

Reply:
Hmm, would something like this be valid?
```
{
    "sobel_cv":
    {
        "ksize": 5
    },
    "negate":
    {
    },
    "order": ["sobel_cv", "negate"]
}
```
or am I abusing the syntax too much? It has the advantage of "fixing" the problem by allowing a precise order to be given, but is backwards compatible if the "order" is omitted, so long as we preserve the order when reading (which we do). Though you could never have a stage actually named "order", but TBH I don't think that bothers me!

Reply:
Yeah, I guess that could work, as long as your parser was updated to treat:
```json
{
    "sobel_cv":
    {
        "ksize": 5
    },
    "negate":
    {
    },
    "order": ["sobel_cv", "negate"]
}
```
and
```json
{
    "negate": {},
    "order": ["sobel_cv", "negate"],
    "sobel_cv": {"ksize": 5}
}
```
as being identical :slightly_smiling_face: 
_Would_ you ever want to do the same operation multiple times (as in my simplistic `["sobel_cv", "negate", "sobel_cv"]` example) ?
I know even less about image-processing than I do about cameras :laughing:

Reply:
Shouldn't the order be moved outside the block, therefor being at the same hierarchical level as the list of stages?  That way its name doesn't clash, and there are parser ordering issues.

Reply:
Possibly, but I don't think the JSON parser (we use the Boost one) will accpet that.

To answer the other question, I don't think we have any stages that we want to run multiple times, though I agree it's not impossible to imagine such a scenario. If that became important we could introduce some syntax that would allow it ("negate.0" and "negate.1" perhaps), but it's not something I feel we need to worry about at the moment.

Reply:
>> Shouldn't the order be moved outside the block, therefor being at the same hierarchical level as the list of stages?

> Possibly, but I don't think the JSON parser (we use the Boost one) will accpet that.

Yeah, the outermost `{  }` is already the top-level hierarchy of the JSON file, so it's impossible to have anything else at the "same level". (If I'm understanding what James is asking for)

Reply:
OK, perhaps you need to add an extra item in each stage which is the position or order of it in the list, i.e. explicitly state it 
 since JSON has no implicit ordering I assume.

`
{
    "sobel_cv":
    {
        "ksize": 5
        "order": 0
    },
    "negate":
    {
        "order": 1
    },
}
`

Reply:
Sounds like you shouldn't be using JSON at all really? Like XML JSON is explicitly non-ordered.

Reply:
OK, thanks for the comments. For now I propose to leave this alone (we've been using JSON files elsewhere for quite a while and treat them in the same way), but adding an "order" field in future seems reasonable as it resolves the issue and is backwards compatible.